### ISSUE TITLE
As a software developer of scientific software, I need a native Haskell implementation of the ddot function so that I can compute the inner product of two vectors in a type-safe, thread safe manner.

### ISSUE THREAD

**Issue by dlewissandy** [dlewissandy/lambda-blas#73]
Documentation for the ddot function can be found at BLAS.

- [x] create native implementation of function
- [x] create haddock documentation of function
- [x] create unit test to demonstrate byte equivalence to FORTRAN BLAS library function
- [x] create benchmark(s) to quantify time performance relative to FORTRAN BLAS library function
- [x] demonstrate that native has a relative time performance of no more than 100x the FORTRAN blas function.

**Comment by dlewissandy** [comment]
Implementation Strategy:
* This function can be implemented easily by polymorphising the single precision version.

**Comment by dlewissandy** [comment]
Attached is the raw benchmark data:
[ddot.zip](https://github.com/dlewissandy/lambda-blas/files/868049/ddot.zip)

**Comment by dlewissandy** [comment]
# DDOT BENCHMARK ANALYSIS #

## Special Case inc=1 ##
When the increment of traversal is equal to unity, the time performance of both the unsafe and pure implementations are well described by a linear model:

### PURE ###
```
lm(formula = PURE ~ N, data = dat[dat$INC == 1, ])

Residuals:
      Min        1Q    Median        3Q       Max
-0.079282  0.002870  0.005404  0.006714  0.049952

Coefficients:
             Estimate Std. Error t value Pr(>|t|)
(Intercept) 1.852e-02  4.398e-03    4.21 0.000157 ***
N           1.010e-03  6.686e-07 1510.84  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.02481 on 37 degrees of freedom
Multiple R-squared:      1,	Adjusted R-squared:      1
F-statistic: 2.283e+06 on 1 and 37 DF,  p-value: < 2.2e-16
```
### UNSAFE ###
```
lm(formula = UNSAFE ~ N, data = dat[dat$INC == 1, ])

Residuals:
      Min        1Q    Median        3Q       Max
-0.051028 -0.005191 -0.004671  0.001203  0.107754

Coefficients:
             Estimate Std. Error  t value Pr(>|t|)
(Intercept) 3.609e-02  4.144e-03    8.708 1.73e-10 ***
N           1.003e-03  6.300e-07 1591.908  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.02338 on 37 degrees of freedom
Multiple R-squared:      1,	Adjusted R-squared:      1
F-statistic: 2.534e+06 on 1 and 37 DF,  p-value: < 2.2e-16
```
### ASYMPTOTIC PERFORMANCE ###
In the limit as n becomes very large, the relative time performance should approach:
```
1.010e-03
--------- = 1.0x
1.003e-03
```
which is within the 100x performance threshold for this project milestone

## Special Case inc/=1 ##
When the increment of traversal is not equal to unity, the time performance of both the unsafe and pure implementations are well described by a linear model:

### PURE ###
```
lm(formula = PURE ~ N, data = dat[dat$INC != 1, ])

Residuals:
     Min       1Q   Median       3Q      Max
-1.00508  0.01635  0.09535  0.13370  0.98187

Coefficients:
              Estimate Std. Error t value Pr(>|t|)
(Intercept) -9.094e-02  4.906e-02  -1.854   0.0698 .
N            5.593e-03  8.486e-05  65.913   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.3196 on 49 degrees of freedom
Multiple R-squared:  0.9888,	Adjusted R-squared:  0.9886
F-statistic:  4344 on 1 and 49 DF,  p-value: < 2.2e-16
```
### UNSAFE ###
```
lm(formula = UNSAFE ~ N, data = dat[dat$INC != 1, ])

Residuals:
     Min       1Q   Median       3Q      Max
-1.24802  0.00628  0.13194  0.18685  1.33907

Coefficients:
              Estimate Std. Error t value Pr(>|t|)
(Intercept) -0.1614408  0.0631318  -2.557   0.0137 *
N            0.0029975  0.0001092  27.451   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.4112 on 49 degrees of freedom
Multiple R-squared:  0.9389,	Adjusted R-squared:  0.9377
F-statistic: 753.5 on 1 and 49 DF,  p-value: < 2.2e-16
```
### ASYMPTOTIC PERFORMANCE ###
In the limit as n becomes very large, the relative time performance should approach:
```
5.593e-03
--------- = 1.9x
2.997e-03
```
which is within the 100x performance threshold for this project milestone.

**Comment by dlewissandy** [comment]
Closed as DONE after approval of pull request #80.
